File: data structures and algorithms/cpp/04_012.cpp
#include <memory>

class ConfigManager {
private:
    static std::shared_ptr<ConfigManager> instance;
    
    // Private constructor to prevent direct instantiation
    ConfigManager() {}
    
public:
    // Prevent copy and assignment
    ConfigManager(const ConfigManager&) = delete;
    ConfigManager& operator=(const ConfigManager&) = delete;
    
    static std::shared_ptr<ConfigManager> getInstance() {
        if (!instance) {
            instance = std::shared_ptr<ConfigManager>(
                new ConfigManager());
        }
        return instance;
    }
};

// Initialize static member
std::shared_ptr<ConfigManager> ConfigManager::instance = nullptr;

File: python/04_012.py
class ConfigManager:
    _instance = None
    
    def __new__(cls):
        if cls._instance is None:
            cls._instance = super(ConfigManager, cls).__new__(cls)
        return cls._instance
    
    @classmethod
    def get_instance(cls):
        if cls._instance is None:
            cls._instance = ConfigManager()
        return cls._instance